frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -4538,71 +4554,96 @@ const CompositeType* isNameOfField(Context* context,
-//    - No valid default-initializer present
-//    - Instantiated generics must supply type/param arguments when
-//      searching for a default-initializer
-//    - Consideration of 'where' clauses
-//    - Composites with compilerError'd default-initializers
-//    - Mutually recursive class types
-//    - Non-nil 'owned' classes
+static const bool&
+canDefaultInitCompositeQuery(Context* context, const CompositeType* ct);
+
+// Determine if a composite type can invoke its default-initializer. TODO...
+//  - Do we need to expand this analysis to consider class types?
+//  - Instantiated generics must supply type/param arguments when
+//    searching for a default-initializer (?).
+//  - Composites with compilerError'd default-initializers.
+//  - Mutually recursive class types.
 static bool
-isTypeDefaultInitializableImpl(Context* context, const Type* t) {
-  const auto g = t->genericity();
+canDefaultInitCompositeImpl(Context* context, const CompositeType* ct) {
+  // It is a type that doesn't have an ID, so we cannot know the answer.
+  if (!ct || ct->id().isEmpty()) return false;
 
-  switch (g) {
-    case Type::CONCRETE: return true;
-    case Type::GENERIC: return false;
+  const auto p = DefaultsPolicy::USE_DEFAULTS;
+  auto& rf = fieldsForTypeDecl(context, ct, p);
 
-    // For these, consider the fields.
-    case Type::GENERIC_WITH_DEFAULTS:
-    case Type::MAYBE_GENERIC:
-      break;
-  }
+  if (rf.isGeneric()) return false;
 
-  CHPL_ASSERT(!t->isPrimitiveType());
-
-  if (t->isBuiltinType()) {
-    CHPL_ASSERT(false && "Not handled!");
-  }
+  bool allFieldsInitializable = true;
+  for (int i = 0; i < rf.numFields(); i++) {
+    auto ft = rf.fieldType(i).type();
 
-  if (auto ct = t->toCompositeType()) {
-    const auto p = DefaultsPolicy::USE_DEFAULTS;
-    auto& rf = fieldsForTypeDecl(context, ct, p);
+    if (auto ct2 = ft->toCompositeType()) {
+      // Recursion of record types is never legal.
+      if (ct2 == ct && ct->isRecordType()) return false;
 
-    if (!rf.isGeneric()) return true;
+      // Bail to avoid an infinite loop.
+      if (context->isQueryRunning(canDefaultInitCompositeQuery, {ct2})) {
+        return false;
+      }
+    }
 
-    // TODO: Do I still need to consider field genericity, here? I.E., if
-    // a field is marked 'GENERIC_WITH_DEFAULTS' is there more to do?
-    // If I can tell the thing is concrete from the ResolvedFields, then
-    // there's probably no need to recurse.
-    if (rf.isGenericWithDefaults()) {
-      for (int i = 0; i < rf.numFields(); i++) {
-        auto ft = rf.fieldType(i).type();
+    // Otherwise, consider the field.
+    if (!rf.fieldHasDefaultValue(i) &&
+        !isTypeWithDefaultValue(context, ft)) {
+      allFieldsInitializable = false;
+      break;
+    }
+  }
 
-        // TODO: Skipping avoids a recursive query but doesn't handle
-        // mutually recursive classes.
-        if (ft == t) continue;
+  const uast::AstNode* ast = nullptr;
```

Comment:
see my above comment -- I'd like us to remove this variable / the use of `idToAst`.

Reply:
Should I plumb this `id` only all the way down into `resolveGeneratedCall` or just stop in the helpers?